File: engines/ruby_engine/engine_server.rb
require 'socket'
require 'json'

PORT = 9008
server = TCPServer.new('127.0.0.1', PORT)
puts "[RUBY ENGINE] Listening on port #{PORT}... (persistent)"

def handle_request(function_name, args)
  case function_name
  when "clean_syntax_process"
    value = args["x"] || ""
    { result: "Ruby processed with clean syntax: #{value}" }
  else
    { error: "unknown function" }
  end
end

def read_exact(socket, num_bytes)
  buffer = ""
  while buffer.length < num_bytes
    chunk = socket.recv(num_bytes - buffer.length)
    break if chunk.nil? || chunk.empty?
    buffer += chunk
  end
  buffer
end

loop do
  client = server.accept

  Thread.new(client) do |conn|
    begin
      header = read_exact(conn, 4)
      length = header.unpack('N')[0]
      body = read_exact(conn, length)

      message = JSON.parse(body)
      function_name = message["function_name"]
      args = message["args"] || {}

      response = handle_request(function_name, args)
      json_response = response.to_json
      response_header = [json_response.bytesize].pack('N')

      conn.write(response_header + json_response)
    rescue => e
      puts "[RUBY ENGINE] Error: #{e.message}"
    ensure
      conn.close
    end
  end
end

File: engines/php_engine/engine_server.php
<?php
$host = '127.0.0.1';
$port = 9007;

$server = stream_socket_server("tcp://$host:$port", $errno, $errstr);
if (!$server) {
    die("[PHP ENGINE] Failed to start: $errstr ($errno)\n");
}

echo "[PHP ENGINE] Listening on port $port... (persistent)\n";

function handle_request($function_name, $args) {
    if ($function_name === "process_web_logic") {
        $value = $args['x'] ?? '';
        return ["result" => "PHP processed web logic: $value"];
    }
    return ["error" => "unknown function"];
}

function read_exact($socket, $num_bytes) {
    $buffer = '';
    while (strlen($buffer) < $num_bytes) {
        $chunk = fread($socket, $num_bytes - strlen($buffer));
        if ($chunk === false || $chunk === '') break;
        $buffer .= $chunk;
    }
    return $buffer;
}

while ($client = stream_socket_accept($server)) {
    // Length-prefixed protocol: 4 bytes header + JSON body
    $header = read_exact($client, 4);
    $length = unpack('N', $header)[1];  // big-endian unsigned int
    $body = read_exact($client, $length);

    $message = json_decode($body, true);
    $function_name = $message['function_name'] ?? '';
    $args = $message['args'] ?? [];

    $response = handle_request($function_name, $args);
    $json_response = json_encode($response);
    $response_header = pack('N', strlen($json_response));

    fwrite($client, $response_header . $json_response);
    fclose($client);
}